OpenRA.Mods.Common/Traits/DockClientBase.cs

```diff
@@ -9,40 +9,63 @@
  */
 #endregion
 
+using System;
 using OpenRA.Primitives;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits
 {
-	public abstract class DockClientBaseInfo : ConditionalTraitInfo, IDockClientInfo, Requires<DockClientManagerInfo> { }
+	public abstract class DockClientBaseInfo : ConditionalTraitInfo, IDockClientInfo, Requires<DockClientManagerInfo>
+	{
+		[Desc("Player relationships the owner of the dock host needs, when not forced enter.",
+			"Use None to only allow the same owner.")]
+		public readonly PlayerRelationship DockRelationships = PlayerRelationship.None;
+
+		[Desc("Player relationships the owner of the dock host needs, when forced enter.",
+			"Use None to only allow the same owner.")]
+		public readonly PlayerRelationship ForceDockRelationships = PlayerRelationship.Ally;
+	}
 
 	public abstract class DockClientBase<InfoType> : ConditionalTrait<InfoType>, IDockClient, INotifyCreated where InfoType : DockClientBaseInfo
 	{
 		readonly Actor self;
 
 		public abstract BitSet<DockType> GetDockType { get; }
 		public DockClientManager DockClientManager { get; }
+		readonly Predicate<Actor> hasDockRelationshipWith;
+		readonly Predicate<Actor> hasDockForcedRelationshipWith;
 
 		protected DockClientBase(Actor self, InfoType info)
 			: base(info)
 		{
 			this.self = self;
 			DockClientManager = self.Trait<DockClientManager>();
+
+			if (info.DockRelationships == PlayerRelationship.None)
+				hasDockRelationshipWith = a => a.Owner == self.Owner;
+			else
+				hasDockRelationshipWith = a => info.DockRelationships.HasRelationship(a.Owner.RelationshipWith(self.Owner));
+
+			if (info.ForceDockRelationships == PlayerRelationship.None)
+				hasDockForcedRelationshipWith = a => a.Owner == self.Owner;
+			else
+				hasDockForcedRelationshipWith = a => info.ForceDockRelationships.HasRelationship(a.Owner.RelationshipWith(self.Owner));
 		}
 
 		protected virtual bool CanDock()
 		{
 			return true;
 		}
 
-		public virtual bool IsDockingPossible(BitSet<DockType> type, bool forceEnter = false)
+		public virtual bool IsDockingPossible(BitSet<DockType> type, bool? forceEnter = false)
 		{
-			return !IsTraitDisabled && GetDockType.Overlaps(type) && (forceEnter || CanDock());
+			return !IsTraitDisabled && GetDockType.Overlaps(type) && (!forceEnter.HasValue || forceEnter.Value || CanDock());
 		}
 
-		public virtual bool CanDockAt(Actor hostActor, IDockHost host, bool forceEnter = false, bool ignoreOccupancy = false)
+		public virtual bool CanDockAt(Actor hostActor, IDockHost host, bool? forceEnter = false, bool ignoreOccupancy = false)
 		{
-			return (forceEnter || self.Owner.IsAlliedWith(hostActor.Owner)) && IsDockingPossible(host.GetDockType, forceEnter) && host.IsDockingPossible(self, this, ignoreOccupancy);
+			return (forceEnter.HasValue ? (forceEnter.Value ? hasDockForcedRelationshipWith(hostActor) : hasDockRelationshipWith(hostActor)) : (hasDockForcedRelationshipWith(hostActor) || hasDockRelationshipWith(hostActor)))
```

Comment:
Why is it necessary to allow null for `forceEnter`? Why not change this to
```suggestion
			return (forceEnter ? (hasDockForcedRelationshipWith(hostActor) || hasDockRelationshipWith(hostActor)) : hasDockRelationshipWith(hostActor))
```

It doesn't make sense to me to disallow force-entering if normal entering would work.

Reply:
Allow null is for `[GenericDockSequence](https://github.com/OpenRA/OpenRA/pull/21067/files#diff-c7420b2896d44687ada8800a0185c3841f47ba9720c6bbd5bb75be74a63dbd55)`. There can be something like this activity that doesn't know that if the call comes from force enter or not, so we use null here, or there will be problem on docking (it will stuck on refinary).

Reply:
In addition, I think we should keep consistent with `TargetRelationship` and `ForceTargetRelationship`, which will be more understandable when using.

Reply:
> There can be something like this activity that doesn't know that if the call comes from force enter or not

Isn't that hiding a bug? If it comes from force entering it should check the extra relations, otherwise not.

> In addition, I think we should keep consistent with `TargetRelationship` and `ForceTargetRelationship` and make enter and force-enter use independent check, which will be more understandable when using.

Look at it from the users point of view: You hover over a unit and get the enter cursor, but as soon as you press the "force enter" hotkey you won't be able to enter. That makes no sense!

Reply:
Our Armament works like this. If `ForceTargetRelationship` doesn't include `TargetRelationship`, then the cursor will be changed to unable attack when you press force fire hot key. Should we consider it flaws, too?

Reply:
> Should we consider Armament is flawed, too?

I would say so.

Reply:
There is application of using the "force attack" operation as ability such as laying mine and spy assassination, when it comes to using `Armament` as 
"force attack" as ability we will also need `ForceTargetRelationship` independent from `TargetRelationship`.

Reply:
That is not a force attack then but a different ability though?

Reply:
If we do need a force enter, I think it should be counted as an ability?